### PR TITLE
Immersive Portals Compatibility

### PR DESCRIPTION
**Pull request by qouteall**
* Removes the only-one-dimension-loaded-at-the-same-time assumption from PipelineManager. (There are still many places in Iris that have this assumption)
* Expose the interface of getting the framebuffer which IP will use
* Add a file to make it buildable by jitpack which IP will use

### CODE REVIEW

File: src/main/java/net/coderbot/iris/pipeline/newshader/NewWorldRenderingPipeline.java [hunk 1]
```diff
@@ -658,4 +658,8 @@ public float getSunPathRotation() {
 	public FrameUpdateNotifier getUpdateNotifier() {
 		return updateNotifier;
 	}
+
+	public GlFramebuffer getBaselineFramebuffer() {
+		return baseline;
+	}
```
Comment (coderbot16) [inline src/main/java/net/coderbot/iris/pipeline/newshader/NewWorldRenderingPipeline.java:664]: See my top-level comment - this isn't needed I don't think
```suggestion
```

**Review by coderbot16 — CHANGES_REQUESTED** [review]
Nice work. I've looked at this PR as well as your [compatibility code in Immersive Portals](https://github.com/qouteall/ImmersivePortalsMod/compare/2eba93abaa97d440b7d3ab2115585672ff01e789...15f00d5fd61d73869ce34e757919d82be3316749).

I designed the architecture of Iris with the existence of mods like Immersive Portals in mind, and clearly, that has paid off. I'm pleased with how relatively straightforward most of the compatibility code is overall.

I have a few comments:

1. First, the manner in which the depth buffer is accessed could be changed. Iris copies the depth buffer content to the main Minecraft framebuffer at the end of WorldRenderer#render so that other mods work as you'd expect (which OptiFine does not appear to do). This means that it's not necessary to access the "baseline" framebuffer to access the depth buffer - instead, you can just access the depth buffer content through the main Minecraft framebuffer once world rendering has completed, that is, `MinecraftClient.getInstance().getFramebuffer()`. (Iris considers world rendering to be completed slightly before Minecraft actually returns from WorldRenderer#render, so you can still inject at the return statement like you appear to already be doing.)

2. Second, the way that you determine whether shaders are currently being used might not always work. In fallback mode, a valid shader pack has been parsed and loaded, but it failed to compile (and therefore cannot be used for rendering), so the rendering pipeline in use will be a FixedFunctionWorldRenderingPipeline (ie, shaders disabled), even though your current code for isShaders will return true in this case. It would be good for me to expose a nice interface just to determine whether shaders are being used for rendering in the immediate moment.

3. Is this code in IrisPortalRenderer needed? I'm curious what behavior is being referenced here, if it's an OptiFine-only thing or if Iris does it too.

```java
    //OptiFine messes up with transformations so store it
    private MatrixStack cachedModelView = new MatrixStack();
```

4. Shader patching / interface: Looks like you patch the shaders that Iris uses, but Iris already has a system for patching shaders. It would likely be preferable for Iris to take responsibility for patching shaders to add the clipping logic and to add / expose / manage the corresponding uniform(s) itself.

5. Pipeline management: The way pipelines are managed in the face of multiple dimensions can have issues. Pipelines for other dimensions are still kept around even if there aren't any portals in sight, and at the same time, multiple portals share the same pipeline. This is problematic for several reasons. In the first case (likely the common case), more memory than necessary will be used, since render targets, compiled shader programs, and other pipeline state takes up a non-trivial amount of memory.

    The second case is also problematic, especially for shaders that make use of data from the previous frame, because if there are multiple portals, the data from the previous frame will be overwritten. For example, the previous camera position will be set to the position of a different portal, and the previous framebuffer content will be set to the viewpoint from a different portal. Naturally, this will result in weird effects, including massive blurring between frames on shader packs that use effects like TAA or motion blur.
    
    The optimal solution is, for each logical viewpoint (player camera + each portal), to store relevant information from the previous frame, and otherwise reuse as much data as possible. Obviously, this is quite complex, requiring some detailed interfaces to be exposed, and I'm not going to hold up the PR over this. However, I would like for the single-dimension behavior to take place if Immersive Portals isn't installed, just to avoid any memory leaks / performance regressions for now for many users using Iris with only Sodium. In the future, I would like to start implementing the more optimal solution, to ensure that bugs are avoided while ensuring that memory is still used efficiently.

In any case, thanks for the PR, and I'm hopeful that in the near future that Iris, Sodium, and Immersive Portals will all be able to run together without significant issues for both good performance and neat portal effects.

**Comment by qouteall** [comment]
The current way of portal rendering with deferred rendering is that, it does not interfere with the deferred rendering process. It renders the outer world and portal world separately and then stitches the final output. This is not the optimal solution because 
1. All the translucent things in front of the portal cannot be rendered correctly.
2. Raindrop distortion does not work with it.
3. The final composition is done multiple times and cost performance.

The optimal way would be:
1. Render the outer world shadow map. After the forward-rendering of the outer world solid things, render the portal inner world. The shadow map should be sampled in the forward pass. The dimension-specific fog color should also be sampled in the forward pass.
2. Then do the forward-rendering of the outer world translucent things.
3. Do only one composition rendering normally.
The reason that I didn't choose the optimal way is that
1. OptiFine allows different shaders for different dimensions, which may use the gbuffer differently
2. The shading of dimension-specific fog color and some other things (such as sampling shadow map) may be deferred, so only one composite rendering won't work.

**Comment by qouteall** [comment]
IP has two modes of portal rendering with shaders(OptiFine or Iris). The first one is "compatibility renderer" which is simpler and faster but does not allow portal-in-portal rendering. The second one allows recursive portal-in-portal rendering but is more complex. It uses one framebuffer per portal layer. I think that it can be improved.

**Comment by qouteall** [comment]
Replying your comments
1. Yes I am thinking simplyfing my portal rendering algorithm using that
4. IP has its own shader patching mechanic which is also required when not using iris.
5. I am going to make that, after a conventional dimension change (with loading screen) it should automatically destroy all pipelines. So that pipeline resource won't leak for non-IP users.
And the logical viewpoint can be discriminated by List<UUID> from IP

File: jitpack.yml [hunk 2]
```diff
@@ -0,0 +1,3 @@
+before_install:
```
[the remark this answers is not in the document]
Comment (qouteall) [reply jitpack.yml:1]: It doesn't work as I tried. I think the current one is fine.

**Review by coderbot16 — APPROVED** [review]
Thanks! Sorry for the delay in getting this merged.

**Pull request merged**